### ISSUE TITLE
PoC Flatpak packaging

### ISSUE THREAD

**Issue by tinywrkb** [virt-manager/virt-manager#246]
I posted about this a while ago in the Red Hat Bugzilla, but as the issue tracker moved to here, I thought maybe it's a good idea to post here.

This is a proof-of-concept Flatpak packaging of Virt Manager. This is not a feature request and I don't plan to submit this to Flathub, or at least without a co-maintainer as I don't use much of the app features, I'm making very limited use of the app, I'm just running a VM locally.  
I hope to get some feedback and hear if something is broken.

Maintainers, feel free to close or lock this issue if you think this is not the right place to discuss it (maybe enable GH's discussions?).

[Link to Flatpak manifest](https://github.com/tinywrkb/flatpaks/blob/master/org.virt_manager.virt-manager/org.virt_manager.virt-manager.yaml)

#### Notes:

* There are a few notes in the manifest, mainly about SSH. The TLDR is that SSH remoting is working.
* I packaged in also Virt Viewer for no good reason other than to save some build time, CPU cycles, and less maintenance. Maybe it should be dropped if this will be submitted to Falthub.
* USB redirection is magically working, at least when I test a USB card reader.  
I can't explain this, I expected that this will require a flatpak-spawn wrapper so usbredirserver or spice-client-glib-usb-acl-helper calls will be made on the host and not in the sandbox.  
I should probably read the code and try to understand this.
* I'm not sure if we need desktop notifications access, PCSC socket, or CUPS socket.
* I recently started to experience some lockup of the Virt Manager GUI. This might be guest performance-related as I really need to upgrade my machine. So this needs to be tested by someone else.

#### Flatpak related Virt Manager issues
* Using a hyphen in the bus name, `org.virt-manager.virt-manager`, is not allowed except the last section of the app-id. If we converted `org.virt-manager.virt-manager` to proper Flatpak app-id then it will be `org.virt_manager.virt-manager`.  
Flatpak automatically allows an app to own a bus name on the user DBus session with the same Flatpak app-id.  
Because the app tries to create instead `org.virt-manager.virt-manager` that is different from the app-id, then I need to add this specific permission to the Flatpak manifest.  
It would be prefered that Virt Manager will create bus name that can be used as Flatpak app-id.  
I recommend to use this chance to follow the more strict [interface name convention](https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names) and rename the bus name to `org.virt_manager.VirtManager`.
* Desktop file and icons should use the reverse DNS naming convention to be Flatpak exportable. So I expect something like `org.virt_manager.VirtManager.desktop`, `org.virt_manager.VirtManager.png`, and `Icon=org.virt_manager.VirtManager` in the desktop file. This should be also [fixed in systray.py](https://github.com/tinywrkb/flatpaks/blob/40b74a645056b2330e3b312f14cd4a4274184c3e/org.virt_manager.virt-manager/virt-manager-systray-icon-use-reverse-dns-naming-convention.patch).

**Comment by crobinso** [comment]
Thanks for working on this! I have no experience with packaging and maintaining flatpaks so I probably will not help here anytime soon.

Thanks for explaining those two issues! I would gladly take patches for those + switching to the interface naming spec

**Comment by ljrk0** [comment]
Since GNOME Boxes uses libvirt as well, maybe one can have a look there for some other things one might re-use:
https://github.com/flathub/org.gnome.Boxes

**Comment by anthr76** [comment]
Hi @tinywrkb any chance you can publish this to flathub?

**Comment by tinywrkb** [comment]
@anthr76, sorry, but I won't be submitting the app to Flathub.  
Publishing an app is a pledge to the quality of its packaging and that all its features are working in different setups.  
I'm already maintaining a couple of applications that their original maintainers thoughtlessly submitted them, without consulting their developers, and without making sure that important features are working correctly.  
I use very little of libvirt/qemu's features, I have a non-standard libvirt installation in a Podman Pod, and the total uptime of my VMs is a handful of hours in a year, so it would be unwise for me to publish the application.

**Comment by anthr76** [comment]
Being a `rpm-ostree` user having a Flatpak is quite valuable for maintaining remote VMs and even VMs locally. Most basic functions are working quite well and infarct and I haven't ran into any issues yet. I may not be the most advanced user but we are able to leverage flathub-beta to gate feedback. I'm happy to pitch in and help. GNOME Boxes has been in flathub for quite a while with various 'quirks'.

Even libvirt socket over ssh with Yubikey PIV is working great!

It would be a shame not to publish this..

**Comment by tinywrkb** [comment]
Well, when I put my maintainer hat on, I try to make sure the application work not only for me, but also for other users.  
I don't want to end up with users opening bug reports on the upstream bugtracker because of broken features in the Flathub app.  

Also, if the developer of the application is not involved in the Flathub app maintenance, then the maintainer becomes responsible to make Flatpak-specific changes to the app. For example, it's likely that we will need to support [the suggested USB portal](https://github.com/flatpak/xdg-desktop-portal/pull/559).  
This is why it's preferable that the developer of the application will maintain its Flathub packaging.

**Comment by tinywrkb** [comment]
I updated the packaging for the 4.0.0 release and all the dependencies.

**Comment by crobinso** [comment]
@tinywrkb thanks for your work. your reasoning makes sense to me. when I have time I will study this some more, I'm new to building flatpaks

**Comment by tinywrkb** [comment]
@crobinso to help a little I can

1. Create an example branch of Flathub submission PR with the Flatpak packaging, which will look much more like a submission proposal.  
This would have to be a single manifest, which will be pretty big considering the number of modules/libs, because Flathub admins are not keen on separated manifests for modules.  
The manifest could be in a JSON or YAML format. I prefer YAML, but I've seen at least one Flathub admin commenting for JSON and against YAML.

2. Send a PR here with a Flatpak manifest that could be used to build and package using the local dir as source for virt-manager. If you really like, you could even generate a Flatpak bundle with this (single file Flatpak package), though Flathub distribution is a better choice. If you'll ever choose to distribute bundles, then you should probably disable locale extensions generation (`separate-locales: false`).  
And same as 1., the manifest could be YAML or JSON format, and would be a single file manifest to avoid polluting the repo with packaging-related files.

In either case, proper testing is needed, which I've never done.

**Comment by tinywrkb** [comment]
I started preparing a [branch here](https://github.com/tinywrkb/flathub/tree/org.virt_manager.virt-manager) in accordance with Flathub submission requirements. This would save some time to whoever submit the app.  
Right now, it just merged the different module manifests I had into a single Flatpak YAML manifest, and without other changes.

**Comment by Arcitec** [comment]
That's looking like a very well-made Flathub manifest, with the automatic update checking configured. But I completely understand that you don't want to maintain it and that it's impossible to know if something will be broken.

I expect that there will be much more interest in this later, as immutable distros continue growing in popularity.

**Comment by Francewhoa** [comment]
Hello Virt-Manager enthusiasts :)

Any volunteers to publish and maintain Virt-Manager to this Flathug repository at https://flathub.org ?

If needed, both me and the Ubertus.Org team would be happy to contribute testing and documentation

--- --- --- --- --- --- --- --- --- --- --- --- --- --- --- ---

For those not familiar with Flathug and Flatpak, below is the same suggestion as above. But with details if you're interested in those.

--- --- --- --- --- --- --- --- --- --- --- --- --- --- --- ---


## Benefit for maintainers:

• All-in-one packages. You save a lot of time. Because Flatpak doesn't depend on outside packages. And Flatpak automatically finds what to bundle using deduplication.

• Single package for all Linux distributions. List of distributions at https://flatpak.org/setup/

• Rapid updates. Send your updates directly to all your users. Your users do not have to wait for package repositories to update their list.

• Strong security. Using automated sandbox. With optional override for users of Flatseal.

• Use any repository to your liking. Your own, the https://codeberg.org , GitLab, GitHub, any others. Details at https://docs.flatpak.org/en/latest/publishing.html

• Developer guide at https://docs.flatpak.org/en/latest/

• Free community support for maintainers: 

___• Forum at https://discourse.flathub.org/

___• Instant message at https://matrix.to/#/#flatpak:matrix.org

___• Mailing list at https://lists.freedesktop.org/mailman/listinfo/flatpak

--- --- --- --- --- --- --- --- --- --- --- --- --- --- --- ---


## Benefit for users:

• Easy and quick installation of the latest free application version at https://flathub.org

• Automatically use any theme to your liking for all apps or selected app. Assuming the app supports theming. This is powered by Flatseal.

___•Steps at https://discourse.flathub.org/t/gtk4-apps-do-not-use-adwaita-dark/1327/2

___•Get Flatseal at https://flathub.org/apps/com.github.tchx84.Flatseal

• Integration with GNOME. You are able to easily install Flatpak applications using the GNOME you're familiar with. When needed, you can still install Flatpak app using https://flathub.org , Terminal, and many other options.

• Of all the packaging options available, Flatpak definitely has the strongest security. Because each application is sanboxed. While at the same time, for some application, the default permissions grant too much access. So users who need strong security need to learn to use Flatseal to adapt the permissions at https://flathub.org/apps/com.github.tchx84.Flatseal

• Freedom. You get more freedom with Flatpak. Because https://flathub.org is controlled by a friendly not-for-profit community of users. In comparison, Snap Store is controlled by the for-profit corporation Canonical. So in the future, under constant pressure from its secret shareholders, when Canonical decides to close Snap Store, you will be in trouble. Not-for-profit community value you the people above profit. There are lots of benefit for you associated with this. Usually, for-profit corporations value money above you, the people. There are lots of risk associated with this.

___• Sources about who controls Flatpak is under "Acknowledgements" title at https://flathub.org/about

___• Sources about who controls Snap Store at https://ubuntu.com/legal/terms-and-policies/snap-store-terms

• Powerful but easy permissions override using this free Flatseal at https://flathub.org/apps/com.github.tchx84.Flatseal

--- --- --- --- --- --- --- --- --- --- --- --- --- --- --- ---


Challenge with Flatpak:

• The first few applications you install use significantly more storage space. Because Flatpak needs to download runtime(s). But those runtimes are downloaded only once and shared with all other applications. This means that on the long term, after installing the first few applications, the storage space used no longer significantly increase.

**Comment by travier** [comment]
Hey folks, I use this application almost daily so I've submitted it to Flathub: https://github.com/flathub/flathub/pull/5882

I'll post patches for the changes that I've made there. As applications on Flathub are ideally co-maintained or co-owned by maintainers, feel free to reach out if you want to be added as a maintainer on Flathub as well. 

Thanks a lot for the initial work @tinywrkb!

**Comment by Francewhoa** [comment]
@travier , @tinywrkb and all, Thanks for your contributions

----------

@travier :) When you are ready, and after it is published on Flathub, both me and the Ubertus.Org team would be happy to contribute testing and documentation. 

>As applications on Flathub are ideally co-maintained or co-owned by maintainers, feel free to reach out if you want to be added as a maintainer on Flathub as well.

For your review for interest and for your decision, if not already done, I suggest that you create a new ticket about your invitation for co-maintainer at https://github.com/flathub/flathub/issues

**Comment by travier** [comment]
The Flatpak is now available on [Flathub](https://flathub.org/apps/org.virt_manager.virt-manager) ([manifest](https://github.com/flathub/org.virt_manager.virt-manager)).

I'm currently maintaining it but if the maintainers are interested, we could make it verified (https://docs.flathub.org/docs/for-app-authors/verification) and get them access to the manifest repo on Flathub as well.

**Comment by travier** [comment]
We have two minor patches for the Flatpak that we should be able to get upstream here:
- https://github.com/flathub/org.virt_manager.virt-manager/blob/master/metadata.patch
- https://github.com/flathub/org.virt_manager.virt-manager/blob/master/virt-manager-systray-icon-use-reverse-dns-naming-convention.patch

**Comment by Francewhoa** [comment]
> The Flatpak is now available on [Flathub](https://flathub.org/apps/org.virt_manager.virt-manager) ([manifest](https://github.com/flathub/org.virt_manager.virt-manager)).

Thanks @travier :) Wow more than 10,000 download in less than 3 months. Congrats.
![Screenshot-557](https://github.com/user-attachments/assets/fb9cb89c-95e7-43e4-8be8-6d1aefdb514e)

**Comment by AsciiWolf** [comment]
Any progress in the [Flatpak verification](https://github.com/flathub/org.virt_manager.virt-manager/issues/2)? :)

**Comment by AsciiWolf** [comment]
> Any progress in the [Flatpak verification](https://github.com/flathub/org.virt_manager.virt-manager/issues/2)? :)

/cc @phrdina
